# Project: Building a Credit Risk Model

In this project, I'll be focusing on credit modelling, a well known data science problem that focuses on modeling a borrower's credit risk. Credit has played a key role in the economy for centuries and some form of credit has existed since the beginning of commerce. I'll be working with financial lending data from Lending Club. Lending Club is a marketplace for personal loans that matches borrowers who are seeking a loan with investors looking to lend money and make a return.

Each borrower fills out a comprehensive application, providing their past financial history, the reason for the loan, and more. Lending Club evaluates each borrower's credit score using past historical data and assign an interest rate to the borrower. The interest rate is the percent in addition to the requested loan amount the borrower has to pay back. Lending Club also tries to verify each piece of information the borrower provides but it can't always verify all of the information (usually for regulation reasons).

A higher interest rate means that the borrower is riskier and more unlikely to pay back the loan while a lower interest rate means that the borrower has a good credit history is more likely to pay back the loan. The interest rates range from 5.32% all the way to 30.99% and each borrower is given a grade according to the interest rate they were assigned. If the borrower accepts the interest rate, then the loan is listed on the Lending Club marketplace.

Investors are primarily interested in receiveing a return on their investments. Approved loans are listed on the Lending Club website, where qualified investors can browse recently approved loans, the borrower's credit score, the purpose for the loan, and other information from the application. Once they're ready to back a loan, they select the amount of money they want to fund. Once a loan's requested amount is fully funded, the borrower receives the money they requested minus the origination fee that Lending Club charges.

The borrower then makes monthly payments back to Lending Club either over 36 months or over 60 months. Lending Club redistributes these payments to the investors. This means that investors don't have to wait until the full amount is paid off to start to see money back. If a loan is fully paid off on time, the investors make a return which corresponds to the interest rate the borrower had to pay in addition the requested amount. Many loans aren't completely paid off on time, however, and some borrowers default on the loan.

While Lending Club has to be extremely savvy and rigorous with their credit modelling, investors on Lending Club need to be equally as savvy about determining which loans are more likely to be paid off. While at first, you may wonder why investors would put money into anything but low interest loans. The incentive investors have to back higher interest loans is, well, the higher interest! If investors believe the borrower can pay back the loan, even if he or she has a weak financial history, then investors can make more money through the larger additional amount the borrower has to pay.

Most investors use a portfolio strategy to invest small amounts in many loans, with healthy mixes of low, medium, and interest loans. In this course, we'll focus on the mindset of a conservative investor who only wants to invest in the loans that have a good chance of being paid off on time. To do that, we'll need to first understand the features in the dataset and then experiment with building machine learning models that reliably predict if a loan will be paid off or not.

# I. Reading in the Dataset: Pandas

In [1]:
import pandas as pd

loans_2007 = pd.read_csv("loans_2007.csv")
loans_2007.head()

/home/ernhung92/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501  1296599.0     5000.0       5000.0           4975.0   36 months   
1  1077430  1314167.0     2500.0       2500.0           2500.0   60 months   
2  1077175  1313524.0     2400.0       2400.0           2400.0   36 months   
3  1076863  1277178.0    10000.0      10000.0          10000.0   36 months   
4  1075358  1311748.0     3000.0       3000.0           3000.0   60 months   

  int_rate  installment grade sub_grade    ...    last_pymnt_amnt  \
0   10.65%       162.87     B        B2    ...             171.62   
1   15.27%        59.83     C        C4    ...             119.66   
2   15.96%        84.33     C        C5    ...             649.91   
3   13.49%       339.31     C        C1    ...             357.48   
4   12.69%        67.79     B        B5    ...              67.79   

  last_credit_pull_d collections_12_mths_ex_med  policy_code application_type  \
0           Jun-2016                        0.0          1.0       INDIVIDUAL   
1           Sep-2013                        0.0          1.0       INDIVIDUAL   
2           Jun-2016                        0.0          1.0       INDIVIDUAL   
3           Apr-2016                        0.0          1.0       INDIVIDUAL   
4           Jun-2016                        0.0          1.0       INDIVIDUAL   

  acc_now_delinq chargeoff_within_12_mths delinq_amnt pub_rec_bankruptcies  \
0            0.0                      0.0         0.0                  0.0   
1            0.0                      0.0         0.0                  0.0   
2            0.0                      0.0         0.0                  0.0   
3            0.0                      0.0         0.0                  0.0   
4            0.0                      0.0         0.0                  0.0   

  tax_liens  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  

[5 rows x 52 columns]

In [2]:
loans_2007.shape

(42538, 52)

The Dataframe contains many columns and can be cumbersome to try to explore all at once. Let's break up the columns into 3 groups of 18 columns and use the data dictionary to become familiar with what each column represents. We  want to pay attention to any features that:
1. leak information from the future (after the loan has already been funded)
2. don't affect a borrower's ability to pay back a loan (e.g. a randomly generated ID value by Lending Club)
3. formatted poorly and need to be cleaned up
4. require more data or a lot of processing to turn into a useful feature
5. contain redundant information

We need to especially pay attention to data leakage, since it can cause our model to overfit. This is because the model would be using data about the target column that wouldn't be available when we're using the model on future loans. 

# II. First Group of Columns

After analyzing the first group of column, these features need to be removed:

1. <b>id</b>: randomly generated field by Lending Club for unique identification purposes only
2. <b>member_id</b>: also a randomly generated field by Lending Club for unique identification purposes only
3. <b>funded_amnt</b>: leaks data from the future (after the loan is already started to be funded)
4. <b>funded_amnt_inv</b>: also leaks data from the future (after the loan is already started to be funded)
5. <b>grade</b>: contains redundant information as the interest rate column (int_rate)
6. <b>sub_grade</b>: also contains redundant information as the interest rate column (int_rate)
7. <b>emp_title</b>: requires other data and a lot of processing to potentially be useful
8. <b>issue_d</b>: leaks data from the future (after the loan is already completed funded)

In [3]:
cols_to_remove = ["id", "member_id", "funded_amnt", "funded_amnt_inv", "grade", "sub_grade", "emp_title", "issue_d"]
loans_2007 = loans_2007.drop(cols_to_remove, axis=1)
loans_2007.head()

loan_amnt        term int_rate  installment emp_length home_ownership  \
0     5000.0   36 months   10.65%       162.87  10+ years           RENT   
1     2500.0   60 months   15.27%        59.83   < 1 year           RENT   
2     2400.0   36 months   15.96%        84.33  10+ years           RENT   
3    10000.0   36 months   13.49%       339.31  10+ years           RENT   
4     3000.0   60 months   12.69%        67.79     1 year           RENT   

   annual_inc verification_status  loan_status pymnt_plan    ...      \
0     24000.0            Verified   Fully Paid          n    ...       
1     30000.0     Source Verified  Charged Off          n    ...       
2     12252.0        Not Verified   Fully Paid          n    ...       
3     49200.0     Source Verified   Fully Paid          n    ...       
4     80000.0     Source Verified      Current          n    ...       

  last_pymnt_amnt last_credit_pull_d collections_12_mths_ex_med policy_code  \
0          171.62           Jun-2016                        0.0         1.0   
1          119.66           Sep-2013                        0.0         1.0   
2          649.91           Jun-2016                        0.0         1.0   
3          357.48           Apr-2016                        0.0         1.0   
4           67.79           Jun-2016                        0.0         1.0   

   application_type  acc_now_delinq chargeoff_within_12_mths  delinq_amnt  \
0        INDIVIDUAL             0.0                      0.0          0.0   
1        INDIVIDUAL             0.0                      0.0          0.0   
2        INDIVIDUAL             0.0                      0.0          0.0   
3        INDIVIDUAL             0.0                      0.0          0.0   
4        INDIVIDUAL             0.0                      0.0          0.0   

   pub_rec_bankruptcies  tax_liens  
0                   0.0        0.0  
1                   0.0        0.0  
2                   0.0        0.0  
3                   0.0        0.0  
4                   0.0        0.0  

[5 rows x 44 columns]

# III. Second Group of Columns

Within this group of columns, we need to drop the following columns:
1. <b>zip_code</b>: redundant with the addr_state column since only the first 3 digits of the 5 digit zip code are visible (which only can be used to identify the state the borrower lives in)
2. <b>out_prncp</b>: leaks data from the future, (after the loan already started to be paid off)
3. <b>out_prncp_inv</b>: also leaks data from the future, (after the loan already started to be paid off)
4. <b>total_pymnt</b>: also leaks data from the future, (after the loan already started to be paid off)
5. <b>total_pymnt_inv</b>: also leaks data from the future, (after the loan already started to be paid off)
6. <b>total_rec_prncp</b>: also leaks data from the future, (after the loan already started to be paid off)

The out_prncp and out_prncp_inv both describe the outstanding principal amount for a loan, which is the remaining amount the borrower still owes. These 2 columns as well as the total_pymnt column describe properties of the loan after it's fully funded and started to be paid off. This information isn't available to an investor before the loan is fully funded and we don't want to include it in our model.

In [4]:
cols_to_drop2 = ["zip_code", "out_prncp", "out_prncp_inv", "total_pymnt", "total_pymnt_inv", "total_rec_prncp"]
loans_2007 = loans_2007.drop(cols_to_drop2, axis=1)
loans_2007.head()

loan_amnt        term int_rate  installment emp_length home_ownership  \
0     5000.0   36 months   10.65%       162.87  10+ years           RENT   
1     2500.0   60 months   15.27%        59.83   < 1 year           RENT   
2     2400.0   36 months   15.96%        84.33  10+ years           RENT   
3    10000.0   36 months   13.49%       339.31  10+ years           RENT   
4     3000.0   60 months   12.69%        67.79     1 year           RENT   

   annual_inc verification_status  loan_status pymnt_plan    ...      \
0     24000.0            Verified   Fully Paid          n    ...       
1     30000.0     Source Verified  Charged Off          n    ...       
2     12252.0        Not Verified   Fully Paid          n    ...       
3     49200.0     Source Verified   Fully Paid          n    ...       
4     80000.0     Source Verified      Current          n    ...       

  last_pymnt_amnt last_credit_pull_d collections_12_mths_ex_med  policy_code  \
0          171.62           Jun-2016                        0.0          1.0   
1          119.66           Sep-2013                        0.0          1.0   
2          649.91           Jun-2016                        0.0          1.0   
3          357.48           Apr-2016                        0.0          1.0   
4           67.79           Jun-2016                        0.0          1.0   

   application_type acc_now_delinq  chargeoff_within_12_mths  delinq_amnt  \
0        INDIVIDUAL            0.0                       0.0          0.0   
1        INDIVIDUAL            0.0                       0.0          0.0   
2        INDIVIDUAL            0.0                       0.0          0.0   
3        INDIVIDUAL            0.0                       0.0          0.0   
4        INDIVIDUAL            0.0                       0.0          0.0   

   pub_rec_bankruptcies  tax_liens  
0                   0.0        0.0  
1                   0.0        0.0  
2                   0.0        0.0  
3                   0.0        0.0  
4                   0.0        0.0  

[5 rows x 38 columns]

# IV. Third Group of Features 

In the last group of columns, we need to drop the following columns:
1. <b>total_rec_int</b>: leaks data from the future, (after the loan already started to be paid off),
2. <b>total_rec_late_fee</b>: also leaks data from the future, (after the loan already started to be paid off),
3. <b>recoveries</b>: also leaks data from the future, (after the loan already started to be paid off),
4. <b>collection_recovery_fee</b>: also leaks data from the future, (after the loan already started to be paid off),
5. <b>last_pymnt_d</b>: also leaks data from the future, (after the loan already started to be paid off),
6. <b>last_pymnt_amnt</b>: also leaks data from the future, (after the loan already started to be paid off).

All of these columns leak data from the future, meaning that they're describing aspects of the loan after it's already been fully funded and started to be paid off by the borrower.

In [5]:
cols_to_drop3 = ["total_rec_int", "total_rec_late_fee", "recoveries", "collection_recovery_fee", "last_pymnt_d", "last_pymnt_amnt"]
loans_2007 = loans_2007.drop(cols_to_drop3, axis=1)
loans_2007.head()

loan_amnt        term int_rate  installment emp_length home_ownership  \
0     5000.0   36 months   10.65%       162.87  10+ years           RENT   
1     2500.0   60 months   15.27%        59.83   < 1 year           RENT   
2     2400.0   36 months   15.96%        84.33  10+ years           RENT   
3    10000.0   36 months   13.49%       339.31  10+ years           RENT   
4     3000.0   60 months   12.69%        67.79     1 year           RENT   

   annual_inc verification_status  loan_status pymnt_plan    ...      \
0     24000.0            Verified   Fully Paid          n    ...       
1     30000.0     Source Verified  Charged Off          n    ...       
2     12252.0        Not Verified   Fully Paid          n    ...       
3     49200.0     Source Verified   Fully Paid          n    ...       
4     80000.0     Source Verified      Current          n    ...       

  initial_list_status last_credit_pull_d collections_12_mths_ex_med  \
0                   f           Jun-2016                        0.0   
1                   f           Sep-2013                        0.0   
2                   f           Jun-2016                        0.0   
3                   f           Apr-2016                        0.0   
4                   f           Jun-2016                        0.0   

   policy_code  application_type acc_now_delinq  chargeoff_within_12_mths  \
0          1.0        INDIVIDUAL            0.0                       0.0   
1          1.0        INDIVIDUAL            0.0                       0.0   
2          1.0        INDIVIDUAL            0.0                       0.0   
3          1.0        INDIVIDUAL            0.0                       0.0   
4          1.0        INDIVIDUAL            0.0                       0.0   

   delinq_amnt  pub_rec_bankruptcies  tax_liens  
0          0.0                   0.0        0.0  
1          0.0                   0.0        0.0  
2          0.0                   0.0        0.0  
3          0.0                   0.0        0.0  
4          0.0                   0.0        0.0  

[5 rows x 32 columns]

In [6]:
loans_2007.shape

(42538, 32)

# V. Target Column

We should use the <b>loan_status</b> column, since it's the only column that directly describes if a loan was paid off on time, had delayed payments, or was defaulted on the borrower. Currently, this column contains text values and we need to convert it to a numerical one for training a model. Let's explore the different values in this column and come up with a strategy for converting the values in this column.

In [7]:
print(loans_2007["loan_status"].value_counts())

Fully Paid                                             33136
Charged Off                                             5634
Does not meet the credit policy. Status:Fully Paid      1988
Current                                                  961
Does not meet the credit policy. Status:Charged Off      761
Late (31-120 days)                                        24
In Grace Period                                           20
Late (16-30 days)                                          8
Default                                                    3
Name: loan_status, dtype: int64


There are 9 different possible values for the <b>loan_status</b> column

Here are the explanations:
1. 33136 loans have been fully paid off
2. 5634 loans for which there are no longer a reasonable expectation of further payments
3. 1988 loans that were paid off but the loan application today would no longer meet the credit policy and wouldn't be approved on to the marketplace
4. 761 loans that were charged of but the loan application today would no longer meet the credit policy and wouldn't be approved on to the marketplace
5. 20 loans are past due but still in the grace period of 15 days
6. 8 loans haven't been paid in 16 to 30 days
7. 24 loans that haven't been paid in 31 to 120 days
8. 961 loans are up to date on current payments
9. 3 loans are defaulted on and no payment has been made for more than 121 days

From the investor's perspective, we're interested in trying to predict which loans will be paid off on time and which ones won't be. Only the Fully Paid and Charged Off values describe the final outcome of the loan. The other values describe loans that are still on going and where the jury is still out on if the borrower will pay back the loan on time or not. While the Default status resembles the Charged Off status, in Lending Club's eyes, loans that are charged off have essentially no chance of being repaid while default ones have a small chance

# VI. Binary Classification

Since we're interested in being able to predict which of these 2 values a loan will fall under, we can treat the problem as a binary classification one. Let's remove all the loans that don't contain either Fully Paid and Charged Off as the loan's status and then transform the Fully Paid values to 1 for the positive case and the Charged Off values to 0 for the negative case

Lastly, one thing we need to keep in mind is the class imbalance between the positive and negative cases. While there are 33,136 loans that have been fully paid off, there are only 5,634 that were charged off. This class imbalance is a common problem in binary classification and during training, the model ends up having a strong bias towards predicting the class with more observations in the training set and will rarely predict the class with less observations. The stronger the imbalance, the more biased the model becomes.

In [8]:
loans_2007 = loans_2007[(loans_2007['loan_status'] == "Fully Paid") | (loans_2007['loan_status'] == "Charged Off")]

status_replace = {
    "loan_status" : {
        "Fully Paid": 1,
        "Charged Off": 0,
    }
}

loans_2007 = loans_2007.replace(status_replace)
loans_2007.head()

loan_amnt        term int_rate  installment emp_length home_ownership  \
0     5000.0   36 months   10.65%       162.87  10+ years           RENT   
1     2500.0   60 months   15.27%        59.83   < 1 year           RENT   
2     2400.0   36 months   15.96%        84.33  10+ years           RENT   
3    10000.0   36 months   13.49%       339.31  10+ years           RENT   
5     5000.0   36 months    7.90%       156.46    3 years           RENT   

   annual_inc verification_status  loan_status pymnt_plan    ...      \
0     24000.0            Verified            1          n    ...       
1     30000.0     Source Verified            0          n    ...       
2     12252.0        Not Verified            1          n    ...       
3     49200.0     Source Verified            1          n    ...       
5     36000.0     Source Verified            1          n    ...       

  initial_list_status last_credit_pull_d collections_12_mths_ex_med  \
0                   f           Jun-2016                        0.0   
1                   f           Sep-2013                        0.0   
2                   f           Jun-2016                        0.0   
3                   f           Apr-2016                        0.0   
5                   f           Jan-2016                        0.0   

   policy_code  application_type acc_now_delinq  chargeoff_within_12_mths  \
0          1.0        INDIVIDUAL            0.0                       0.0   
1          1.0        INDIVIDUAL            0.0                       0.0   
2          1.0        INDIVIDUAL            0.0                       0.0   
3          1.0        INDIVIDUAL            0.0                       0.0   
5          1.0        INDIVIDUAL            0.0                       0.0   

   delinq_amnt  pub_rec_bankruptcies  tax_liens  
0          0.0                   0.0        0.0  
1          0.0                   0.0        0.0  
2          0.0                   0.0        0.0  
3          0.0                   0.0        0.0  
5          0.0                   0.0        0.0  

[5 rows x 32 columns]

# V. Removing Single Value Columns

let's look for any columns that contain only one unique value and remove them. These columns won't be useful for the model since they don't add any information to each loan application. In addition, removing these columns will reduce the number of columns we'll need to explore further

In [9]:
cols = loans_2007.columns
drop_columns = []
for col in cols:
    col_series = loans_2007[col].dropna().unique()
    if len(col_series) == 1:
        drop_columns.append(col)
loans_2007 = loans_2007.drop(drop_columns, axis=1)
print(drop_columns)

['pymnt_plan', 'initial_list_status', 'collections_12_mths_ex_med', 'policy_code', 'application_type', 'acc_now_delinq', 'chargeoff_within_12_mths', 'delinq_amnt', 'tax_liens']


In [10]:
loans_2007.head()

loan_amnt        term int_rate  installment emp_length home_ownership  \
0     5000.0   36 months   10.65%       162.87  10+ years           RENT   
1     2500.0   60 months   15.27%        59.83   < 1 year           RENT   
2     2400.0   36 months   15.96%        84.33  10+ years           RENT   
3    10000.0   36 months   13.49%       339.31  10+ years           RENT   
5     5000.0   36 months    7.90%       156.46    3 years           RENT   

   annual_inc verification_status  loan_status         purpose  \
0     24000.0            Verified            1     credit_card   
1     30000.0     Source Verified            0             car   
2     12252.0        Not Verified            1  small_business   
3     49200.0     Source Verified            1           other   
5     36000.0     Source Verified            1         wedding   

          ...          delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  \
0         ...                  0.0         Jan-1985             1.0       3.0   
1         ...                  0.0         Apr-1999             5.0       3.0   
2         ...                  0.0         Nov-2001             2.0       2.0   
3         ...                  0.0         Feb-1996             1.0      10.0   
5         ...                  0.0         Nov-2004             3.0       9.0   

  pub_rec  revol_bal  revol_util  total_acc  last_credit_pull_d  \
0     0.0    13648.0       83.7%        9.0            Jun-2016   
1     0.0     1687.0        9.4%        4.0            Sep-2013   
2     0.0     2956.0       98.5%       10.0            Jun-2016   
3     0.0     5598.0         21%       37.0            Apr-2016   
5     0.0     7963.0       28.3%       12.0            Jan-2016   

  pub_rec_bankruptcies  
0                  0.0  
1                  0.0  
2                  0.0  
3                  0.0  
5                  0.0  

[5 rows x 23 columns]

It looks we we were able to remove 9 more columns since they only contained 1 unique value.

In [11]:
loans_2007.to_csv("filtered_loans_2007.csv", index=False)

This will load the dataframe into a csv file with all our changes

# Preparing the Features

In this part, I'll be preparing the data for machine learning by focusing in handling missing values, converting categorical columns to numeric columns and removing any other extraneous columns

In [12]:
loans = pd.read_csv("filtered_loans_2007.csv")
null_counts = loans.isnull().sum()
print(null_counts)

loan_amnt                 0
term                      0
int_rate                  0
installment               0
emp_length                0
home_ownership            0
annual_inc                0
verification_status       0
loan_status               0
purpose                   0
title                    10
addr_state                0
dti                       0
delinq_2yrs               0
earliest_cr_line          0
inq_last_6mths            0
open_acc                  0
pub_rec                   0
revol_bal                 0
revol_util               50
total_acc                 0
last_credit_pull_d        2
pub_rec_bankruptcies    697
dtype: int64


There are 4 columns with missing values

# I. Handling Missing Values

In [13]:
loans = loans.drop("pub_rec_bankruptcies", axis=1)
loans = loans.dropna()
loans.dtypes.value_counts()

object     11
float64    10
int64       1
dtype: int64

There are some columns that are texts that need to be converted to numerical data types. 

# II. Text Columns

In [14]:
object_columns_df = loans.select_dtypes(include=["object"])
object_columns_df.head()

term int_rate emp_length home_ownership verification_status  \
0   36 months   10.65%  10+ years           RENT            Verified   
1   60 months   15.27%   < 1 year           RENT     Source Verified   
2   36 months   15.96%  10+ years           RENT        Not Verified   
3   36 months   13.49%  10+ years           RENT     Source Verified   
4   36 months    7.90%    3 years           RENT     Source Verified   

          purpose                                  title addr_state  \
0     credit_card                               Computer         AZ   
1             car                                   bike         GA   
2  small_business                   real estate business         IL   
3           other                               personel         CA   
4         wedding  My wedding loan I promise to pay back         AZ   

  earliest_cr_line revol_util last_credit_pull_d  
0         Jan-1985      83.7%           Jun-2016  
1         Apr-1999       9.4%           Sep-2013  
2         Nov-2001      98.5%           Jun-2016  
3         Feb-1996        21%           Apr-2016  
4         Nov-2004      28.3%           Jan-2016

# III. Converting Text Columns

Some of the columns seem like they represent categorical values, but we should confirm by checking the number of unique values in those columns:
1. <b>home_ownership</b>: home ownership status, can only be 1 of 4 categorical values according to the data dictionary,
2. <b>verification_status</b>: indicates if income was verified by Lending Club,
3. <b>emp_length</b>: number of years the borrower was employed upon time of application,
4. <b>term</b>: number of payments on the loan, either 36 or 60,
5. <b>addr_state</b>: borrower's state of residence,
6. <b>purpose</b>: a category provided by the borrower for the loan request,
7. <b>title</b>: loan title provided the borrower,

There are also some columns that represent numeric values, that need to be converted:
1. <b>int_rate</b>: interest rate of the loan in %,
2. <b>revol_util</b>: revolving line utilization rate or the amount of credit the borrower is using relative to all available credit, read more here.

Based on the first row's values for purpose and title, it seems like these columns could reflect the same information. Let's explore the unique value counts separately to confirm if this is true.

Lastly, some of the columns contain date values that would require a good amount of feature engineering for them to be potentially useful:
1. <b>earliest_cr_line</b>: The month the borrower's earliest reported credit line was opened,
2. <b>last_credit_pull_d</b>: The most recent month Lending Club pulled credit for this loan.

Since these date features require some feature engineering for modeling purposes, let's remove these date columns from the Dataframe.

In [15]:
cols = ['home_ownership', 'verification_status', 'emp_length', 'term', 'addr_state']

for c in cols:
    print(loans[c].value_counts())

RENT        18513
MORTGAGE    17112
OWN          2984
OTHER          96
NONE            3
Name: home_ownership, dtype: int64
Not Verified       16696
Verified           12290
Source Verified     9722
Name: verification_status, dtype: int64
10+ years    8545
< 1 year     4513
2 years      4303
3 years      4022
4 years      3353
5 years      3202
1 year       3176
6 years      2177
7 years      1714
8 years      1442
9 years      1229
n/a          1032
Name: emp_length, dtype: int64
 36 months    29041
 60 months     9667
Name: term, dtype: int64
CA    6958
NY    3713
FL    2791
TX    2667
NJ    1798
IL    1483
PA    1473
VA    1376
GA    1364
MA    1301
OH    1179
MD    1026
AZ     850
WA     822
CO     770
NC     753
CT     730
MI     712
MO     671
MN     603
NV     481
SC     462
WI     441
AL     437
OR     436
LA     430
KY     315
OK     290
KS     260
UT     254
AR     237
DC     209
RI     196
NM     184
WV     172
NH     166
HI     166
DE     113
MT      83
WY      80
AK      

the <b>home_ownership</b>, <b>verification_status</b>, <b>emp_length</b>, <b>term</b>, <b>addr_state</b> columns all contain multiple discrete values. <b>emp_length</b> needs to be cleaned up a bit though.

In [16]:
print(loans["purpose"].value_counts())
print(loans["title"].value_counts())

debt_consolidation    18130
credit_card            5039
other                  3864
home_improvement       2897
major_purchase         2155
small_business         1762
car                    1510
wedding                 929
medical                 680
moving                  576
vacation                375
house                   369
educational             320
renewable_energy        102
Name: purpose, dtype: int64
Debt Consolidation                              2104
Debt Consolidation Loan                         1632
Personal Loan                                    642
Consolidation                                    494
debt consolidation                               485
Credit Card Consolidation                        353
Home Improvement                                 346
Debt consolidation                               324
Small Business Loan                              310
Credit Card Loan                                 305
Personal                                         3

The <b>home_ownership</b>, <b>verification_status</b>, <b>emp_length</b>, and <b>term</b> columns each contain a few discrete categorical values. We should encode these columns as dummy variables and keep them.

It seems like the <b>purpose</b> and <b>title</b> columns do contain overlapping information but we'll keep the purpose column since it contains a few discrete values. In addition, the <b>title</b> column has data quality issues since many of the values are repeated with slight modifications (e.g. Debt Consolidation and Debt Consolidation Loan and debt consolidation).

Also, the <b>emp_length</b> column needs to be modified

Lastly, the <b>addr_state</b> contains too many discrete values and adding them as dummy variables could lead to a slow code run. Let's remove it

In [17]:
mapping_dict = {
    "emp_length": {
        "10+ years": 10,
        "9 years": 9,
        "8 years": 8,
        "7 years": 7,
        "6 years": 6,
        "5 years": 5,
        "4 years": 4,
        "3 years": 3,
        "2 years": 2,
        "1 year": 1,
        "< 1 year": 0,
        "n/a": 0
    }
}


loans = loans.drop(["last_credit_pull_d","addr_state","title","earliest_cr_line"], axis=1)
loans["int_rate"] = loans["int_rate"].str.rstrip('%').astype("float")
loans["revol_util"] = loans["revol_util"].str.rstrip('%').astype("float")
loans = loans.replace(mapping_dict)

In [18]:
loans.head()

loan_amnt        term  int_rate  installment  emp_length home_ownership  \
0     5000.0   36 months     10.65       162.87          10           RENT   
1     2500.0   60 months     15.27        59.83           0           RENT   
2     2400.0   36 months     15.96        84.33          10           RENT   
3    10000.0   36 months     13.49       339.31          10           RENT   
4     5000.0   36 months      7.90       156.46           3           RENT   

   annual_inc verification_status  loan_status         purpose    dti  \
0     24000.0            Verified            1     credit_card  27.65   
1     30000.0     Source Verified            0             car   1.00   
2     12252.0        Not Verified            1  small_business   8.72   
3     49200.0     Source Verified            1           other  20.00   
4     36000.0     Source Verified            1         wedding  11.20   

   delinq_2yrs  inq_last_6mths  open_acc  pub_rec  revol_bal  revol_util  \
0          0.0             1.0       3.0      0.0    13648.0        83.7   
1          0.0             5.0       3.0      0.0     1687.0         9.4   
2          0.0             2.0       2.0      0.0     2956.0        98.5   
3          0.0             1.0      10.0      0.0     5598.0        21.0   
4          0.0             3.0       9.0      0.0     7963.0        28.3   

   total_acc  
0        9.0  
1        4.0  
2       10.0  
3       37.0  
4       12.0

# IV. Dummy Variables

In [19]:
cols = ["home_ownership","verification_status","purpose","term"]
dummy_df = pd.get_dummies(loans[cols])
loans = pd.concat([loans, dummy_df], axis=1)
loans = loans.drop(cols, axis=1)

In [20]:
loans.head()

loan_amnt  int_rate  installment  emp_length  annual_inc  loan_status  \
0     5000.0     10.65       162.87          10     24000.0            1   
1     2500.0     15.27        59.83           0     30000.0            0   
2     2400.0     15.96        84.33          10     12252.0            1   
3    10000.0     13.49       339.31          10     49200.0            1   
4     5000.0      7.90       156.46           3     36000.0            1   

     dti  delinq_2yrs  inq_last_6mths  open_acc       ...         \
0  27.65          0.0             1.0       3.0       ...          
1   1.00          0.0             5.0       3.0       ...          
2   8.72          0.0             2.0       2.0       ...          
3  20.00          0.0             1.0      10.0       ...          
4  11.20          0.0             3.0       9.0       ...          

   purpose_major_purchase  purpose_medical  purpose_moving  purpose_other  \
0                       0                0               0              0   
1                       0                0               0              0   
2                       0                0               0              0   
3                       0                0               0              1   
4                       0                0               0              0   

   purpose_renewable_energy  purpose_small_business  purpose_vacation  \
0                         0                       0                 0   
1                         0                       0                 0   
2                         0                       1                 0   
3                         0                       0                 0   
4                         0                       0                 0   

   purpose_wedding  term_ 36 months  term_ 60 months  
0                0                1                0  
1                0                0                1  
2                0                1                0  
3                0                1                0  
4                1                1                0  

[5 rows x 38 columns]

In [21]:
loans.to_csv("cleaned_loans_2007.csv", index=False)

# Making Predictions

In [22]:
loans = pd.read_csv("cleaned_loans_2007.csv")
loans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38708 entries, 0 to 38707
Data columns (total 38 columns):
loan_amnt                              38708 non-null float64
int_rate                               38708 non-null float64
installment                            38708 non-null float64
emp_length                             38708 non-null int64
annual_inc                             38708 non-null float64
loan_status                            38708 non-null int64
dti                                    38708 non-null float64
delinq_2yrs                            38708 non-null float64
inq_last_6mths                         38708 non-null float64
open_acc                               38708 non-null float64
pub_rec                                38708 non-null float64
revol_bal                              38708 non-null float64
revol_util                             38708 non-null float64
total_acc                              38708 non-null float64
home_ownership_MORTGAGE    

# I. Picking an Error Metric

Before diving in and selecting an algorithm to apply to the data, we should select an error metric.

An error metric will help us figure out when our model is performing well, and when it's performing poorly. To tie error metrics all the way back to the original question we wanted to answer, let's say we're using a machine learning model to predict whether or not we should fund a loan on the Lending Club platform. Our objective in this is to make money -- we want to fund enough loans that are paid off on time to offset our losses from loans that aren't paid off. An error metric will help us determine if our algorithm will make us money or lose us money.

In this case, we're primarily concerned with false positives and false negatives. Both of these are different types of misclassifications. With a false positive, we predict that a loan will be paid off on time, but it actually isn't. This costs us money, since we fund loans that lose us money. With a false negative, we predict that a loan won't be paid off on time, but it actually would be paid off on time. This loses us potential money, since we didn't fund a loan that actually would have been paid off.

Since we're viewing this problem from the standpoint of a conservative investor, we need to treat false positives differently than false negatives. A conservative investor would want to minimize risk, and avoid false positives as much as possible. They'd be more okay with missing out on opportunities (false negatives) than they would be with funding a risky loan (false positives).

# II. Logistic Regression

In [23]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()

train_cols = loans.columns.drop("loan_status")
features = loans[train_cols]
target = loans["loan_status"]

lr.fit(features, target)
predictions = lr.predict(features)
print(predictions)

[1 1 1 ..., 1 1 1]


While we generated predictions in here, those predictions were overfit. They were overfit because we generated predictions using the same data that we trained our model on. When we use this to evaluate error, we get an unrealistically high depiction of how accurate the algorithm is, because it already knows the correct answers

# III. Cross Validation

In [24]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_predict
lr = LogisticRegression()

predictions = cross_val_predict(lr, features, target, cv=3)
predictions = pd.Series(predictions)

# False positives.
fp_filter = (predictions == 1) & (loans["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)

print(tpr)
print(fpr)

0.9989121566494424
0.9967943009795192


In [25]:
predictions.head(10)

0    1
1    1
2    1
3    1
4    1
5    1
6    1
7    1
8    1
9    1
dtype: int64

Unfortunately, even through we're not using accuracy as an error metric, the classifier is, and it isn't accounting for the imbalance in the classes

# IV. Penalizing the Classifier

There are a few ways to get a classifier to correct for imbalanced classes. The two main ways are:
1. Use oversampling and undersampling to ensure that the classifier gets input that has a balanced number of each class.
2. Tell the classifier to penalize misclassifications of the less prevalent class more than the other class.


In [26]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_predict
lr = LogisticRegression(class_weight="balanced")

predictions = cross_val_predict(lr, features, target, cv=3)
predictions = pd.Series(predictions)

# False positives.
fp_filter = (predictions == 1) & (loans["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)

print(tpr)
print(fpr)

0.6631916115190524
0.3861086375779163


We significantly improved false positive rate by balancing the classes, which reduced true positive rate. Our true positive rate is now around 67%, and our false positive rate is around 40%. From a conservative investor's standpoint, it's reassuring that the false positive rate is lower because it means that we'll be able to do a better job at avoiding bad loans than if we funded everything. However, we'd only ever decide to fund 67% of the total loans (true positive rate), so we'd immediately reject a good amount of loans.

We can try to lower the false positive rate further by assigning a harsher penalty for misclassifying the negative class. While setting class_weight to balanced will automatically set a penalty based on the number of 1s and 0s in the column, we can also set a manual penalty. In the last screen, the penalty scikit-learn imposed for misclassifying a 0 would have been around 5.89 (since there are 5.89 times as many 1s as 0s)

In [27]:
penalty = {
    0: 10,
    1: 1
}

lr = LogisticRegression(class_weight=penalty)

predictions = cross_val_predict(lr, features, target, cv=3)
predictions = pd.Series(predictions)

# False positives.
fp_filter = (predictions == 1) & (loans["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)

print(tpr)
print(fpr)

0.2359411355875865
0.08744434550311665


It looks like assigning manual penalties lowered the false positive rate to 9%, and thus lowered our risk. Note that this comes at the expense of true positive rate. While we have fewer false positives, we're also missing opportunities to fund more loans and potentially make more money. Given that we're approaching this as a conservative investor, this strategy makes sense

# V. Random Forests

In [28]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.cross_validation import cross_val_predict

lr = RandomForestClassifier(random_state=1,class_weight="balanced")

predictions = cross_val_predict(lr, features, target, cv=3)
predictions = pd.Series(predictions)

# False positives.
fp_filter = (predictions == 1) & (loans["loan_status"] == 0)
fp = len(predictions[fp_filter])

# True positives.
tp_filter = (predictions == 1) & (loans["loan_status"] == 1)
tp = len(predictions[tp_filter])

# False negatives.
fn_filter = (predictions == 0) & (loans["loan_status"] == 1)
fn = len(predictions[fn_filter])

# True negatives
tn_filter = (predictions == 0) & (loans["loan_status"] == 0)
tn = len(predictions[tn_filter])

# Rates
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)

print(tpr)
print(fpr)

/home/ernhung92/anaconda3/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


0.9708699725017376
0.9271593944790739


Unfortunately, using a random forest classifier didn't improve our false positive rate. The model is likely weighting too heavily on the 1 class, and still mostly predicting 1s. We could fix this by applying a harsher penalty for misclassifications of 0s.

Ultimately, our best model had a false positive rate of 9%, and a true positive rate of 24%. For a conservative investor, this means that they make money as long as the interest rate is high enough to offset the losses from 9% of borrowers defaulting, and that the pool of 24% of borrowers is large enough to make enough interest money to offset the losses.

If we had randomly picked loans to fund, borrowers would have defaulted on 14.5% of them, and our model is better than that, although we're excluding more loans than a random strategy would